torch/_inductor/codegen/wrapper.py

```diff
@@ -774,7 +774,21 @@ def generate_kernel_call(
                 stream_ptr = f"c_void_p({stream_name})"
                 self.writeline(f"{name}.{name}({call_args_str}, {stream_ptr})")
         else:
-            self.writeline(self.wrap_kernel_call(name, call_args))
+            if V.graph.aot_mode and config.aot_inductor.abi_compatible:
+                from .cpp import DTYPE_TO_CPP
+                new_args = []
+                for arg in call_args:
+                    var_name = f"var_{next(self.arg_var_id)}"
+                    self.writeline(f"void *{var_name}{self.ending}")
+                    self.writeline(
+                        f"AOTI_TORCH_ERROR_CHECK(aoti_torch_get_data_ptr(&{var_name}, {arg}.get()));"
```

Comment:
Why is the `.get()` needed here?  What is the type of arg?

Reply:
```arg``` is a shared_ptr (```std::shared_ptr<AtenTensorOpaque>```).

Reply:
Will shared_ptr be ABI compatible?

Reply:
It's from my last week's ABI change. std::shared_ptr<AtenTensorOpaque> only lives in model.so, and only raw pointers are passed across ABI (https://github.com/pytorch/pytorch/blob/main/torch/csrc/inductor/aoti_torch/c/shim.h).

I have another PR https://github.com/pytorch/pytorch/pull/109606 to clean up so that we can do implicit type conversion instead of calling .get()  (based on Ed's suggestion).

Reply:
I believe shared pointer needs to allocate reference count (via make_shared) on the object, but the object will be owned by by libtorch.so.  Also the destruct will need to live in libtorch.  So I am not convicted this is safe.

Reply:
In https://github.com/pytorch/pytorch/pull/109606, I changed this to use std::unique_ptr. Nevertheless, we still have to answer the ownership question. 

model.so can only call a certain number of APIs to allocate tensors, and those tensors are allocated by libtorch as raw pointers. For those raw pointers, we wrap them with unique_ptr and set the deleter to call `aoti_torch_delete_tensor_object`, so when unique_ptr goes out of scope, the previously returned tensor object will be freed. The question is whether we can make sure all the returned raw pointers are wrapped properly? Right now, we are generating those wrappings in codegen. IIUC, @ezyang had a suggestion to add another mirroring set of APIs in model.so to take uniqe_ptr and call those `aoti_torch_*` functions. That probably will make it safe.

Reply:
std::unique_ptr might be safer, since I don't think it requires any runtime state (that might not be ABI compatible).

Reply:
Agreed. ```std::unique_ptr``` might be better than ```std::shared_ptr``` here.

Reply:
I mean, both std::unique_ptr and std::shared_ptr in the ABI mean you no longer have a C-only ABI. I was hoping we could stick to C-only ABI as this would also make platform upgrades smoother (since C++ ABI is almost definitely getting broken when you upgrade libstdc++).

@jansel The way Bin has designed it is that there is an ownership contract across the ABI, which is that predictor will pass ownership of an opaque handle to the model.so, and it is the model.so's responsibility to dispose of the handle (by calling another stable C ABI deallocation function) on it when it is done. This seems kosher to me. Some of the confusion is that model.so is using a private(!) RAII class to make managing this deallocation process easier; this is also OK because the RAII object is private use only, it never shows up in the public APIs.

Reply:
Yeah, I was kind of expecting to see a C-only ABI here.  I think this is fine.